addon/-private/validated-changeset.ts

```diff
@@ -0,0 +1,933 @@
+    this[ERRORS] = this.setDeep(errors, (<string>key), newError);
+
+    return { value, validation };
+  }
+
+  /**
+   * Creates a snapshot of the changeset's errors and changes.
+   *
+   * @method snapshot
+   */
+  snapshot(): Snapshot {
+    let changes: Changes = this[CHANGES];
+    let errors: Errors<any> = this[ERRORS];
+
+    return {
+      changes: keys(changes).reduce((newObj: Changes, key: keyof Changes) => {
+        newObj[key] = changes[key].value;
+        return newObj;
+      }, {}),
+
+      errors: keys(errors).reduce((newObj: Errors<any>, key: keyof Errors<any>) => {
+        let e = errors[key]
+        newObj[key] = { value: e.value, validation: e.validation };
+        return newObj;
+      }, {}),
+    };
+  }
+
+  /**
+   * Restores a snapshot of changes and errors. This overrides existing
+   * changes and errors.
+   *
+   * @method restore
+   */
+  restore({ changes, errors }: Snapshot): IChangeset {
+    validateNestedObj('snapshot.changes', changes);
+    validateNestedObj('snapshot.errors',  errors);
+
+    let newChanges: Changes = keys(changes).reduce((newObj: Changes, key: keyof Changes) => {
+      newObj[key] = new Change(changes[key]);
+      return newObj;
+    }, {});
+
+    let newErrors: Errors<any> = keys(errors).reduce((newObj: Errors<any>, key: keyof Changes) => {
+      let e: IErr<any> = errors[key];
+      newObj[key] = new Err(e.value, e.validation);
+      return newObj;
+    }, {});
+
+    // @tracked
+    this[CHANGES] = newChanges;
+    // @tracked
+    this[ERRORS] = newErrors;
+
+    this._notifyVirtualProperties();
+    return this;
+  }
+
+  /**
+   * Unlike `Ecto.Changeset.cast`, `cast` will take allowed keys and
+   * remove unwanted keys off of the changeset. For example, this method
+   * can be used to only allow specified changes through prior to saving.
+   *
+   * @method cast
+   */
+  cast(allowed: string[] = []): IChangeset {
+    let changes: Changes = this[CHANGES];
+
+    if (Array.isArray(allowed) && allowed.length === 0) {
+      return this;
+    }
+
+    let changeKeys: string[] = keys(changes);
+    let validKeys = changeKeys.filter((key: string) => allowed.includes(key));
+    let casted = take(changes, validKeys);
+    // @tracked
+    this[CHANGES] = casted;
+    return this;
+  }
+
+  /**
+   * Checks to see if async validator for a given key has not resolved.
+   * If no key is provided it will check to see if any async validator is running.
+   *
+   * @method isValidating
+   */
+  isValidating(key: string | void): boolean {
+    let runningValidations: RunningValidations = this[RUNNING_VALIDATIONS];
+    let ks: string[] = keys(runningValidations);
+    if (key) {
+      return ks.includes(key);
+    }
+    return ks.length > 0;
+  }
+
+  /**
+   * Validates a specific key
+   *
+   * @method _validateKey
+   * @private
+   */
+  _validateKey<T> (
+    key: string,
+    value: T
+  ): Promise<ValidationResult | T | IErr<T>> | T | IErr<T> | ValidationResult {
```

Comment:
I think this is a great candidate to an async function, so you don't need to split whether the validation is sync or async.